serverless: adding an overview on what is/how works the Serverless Agent (#6905)
DataDog/datadog-agent: docs/serverless/README.md

# Serverless Agent

The Serverless Agent is a build of the Datadog Agent capable of running

> in serverless environments, to collect custom metrics, function's logs and soon

Reviewer: Shouldn't `'s` be used for a person or an animal and not for `function`?

> function and the traces of the function execution. Today, its main usage is to

Reviewer 2: ```suggestion
function and the traces of the function execution. Today, it's mainly used to
```

build the Datadog Lambda Extension for AWS Lambda.

## Concepts

> Today, the Serverless Agent is intended to run as a side-car of the serverless function.

Reviewer 2: ```suggestion
The Serverless Agent is intended to run as a side-car of the serverless function.
```

### Validate the presence of the Serverless Agent and order a flush

> The Serverless Agent is running an HTTP server on port 8124.

Reviewer: `is running` => `runs` ?

> With the Datadog Lambda Extension, this HTTP servers is listening for two

Reviewer: `this HTTP server*s* is listen*ing* *for*` => `this HTTP server listens to` ?

different routes:
    - /lambda/hello
    - /lambda/flush

> These routes are not doing anything in specialized for AWS, however, they can be

Reviewer: `These routes are not doing anything in specialized for AWS`
What does it mean?

> These routes are not doing anything in specialized for AWS, however, they can be

Reviewer 2: ```suggestion
```

agree here, we can get rid of this whole sentence because the important info is in the next sentence.

> at some point. That's the reason that if the Serverless Agent has to support a new

Reviewer 2: ```suggestion
If the Serverless Agent needs to be set up to support a different 
```

> serverless environment, it would be important to create different HTTP routes (most

Reviewer 2: ```suggestion
serverless environment, like GCP or Azure App Services Extension, it would be important to create different HTTP routes (most
```
is this what you mean here? if not, please update with examples of what you do mean

Author: I'll use

`If the Serverless Agent needs to be set up to support a different serverless environment, on GCP or Azure for instance, it would be`

if that sounds good to you.

Reviewer 2: yep, perfect. I just wasn't clear here if you meant a completely different service here or a multiple of the same service

likely sharing a big part of the implementation).

#### `Hello` route

No payload.

> This is a discover route: by calling this HTTP route, a function can know if the

Reviewer 2: ```suggestion
This is a discover route: by calling this HTTP route, a function knows if the
```

Serverless Agent is running.

#### `Flush`

No payload.

This route is used to tell the Serverless Agent that it has to immediately flush

> the data it has buffered. This route is blocking and replying only once all the

Reviewer: `This route is blocking and replying` => `This route blocks and replies`?

data has been flushed to the Datadog intake.

### Configuration

The configuration of the Serverless Agent goes through the regular `config` package

> of the Datadog Agent, reading both in a file and in the environment variables.

Reviewer: `reading both in` => `reading both from`

For the Datadog Lambda Extension, some fields are set directly in the environment
at startup for the config package to pick them up. (See `cmd/serverless/main.go` for the
startup process).

### Build

> In order to build a Serverless Agent, please refer to [this README](https://github.com/DataDog/datadog-agent/tree/master/cmd/serverless/README.md).

Reviewer 2: ```suggestion
For more information about building a Serverless Agent, see [this README](https://github.com/DataDog/datadog-agent/tree/master/cmd/serverless/README.md).
```
